examples/mnist/Main.hs

```diff
@@ -0,0 +1,159 @@
+  MLP { layer0 :: Linear dtype inputFeatures hiddenFeatures
+      , layer1 :: Linear dtype hiddenFeatures hiddenFeatures
+      , layer2 :: Linear dtype hiddenFeatures outputFeatures
+      } deriving (Show, Generic)
+
+instance A.Parameterized (MLP dtype inputFeatures outputFeatures hiddenFeatures)
+
+instance (KnownDType dtype, KnownNat inputFeatures, KnownNat outputFeatures, KnownNat hiddenFeatures) => A.Randomizable (MLPSpec dtype inputFeatures outputFeatures hiddenFeatures) (MLP dtype inputFeatures outputFeatures hiddenFeatures) where
+  sample MLPSpec =
+    MLP <$> A.sample LinearSpec <*> A.sample LinearSpec <*> A.sample LinearSpec
+
+mlp
+  :: MLP dtype inputFeatures outputFeatures hiddenFeatures
+  -> Tensor dtype '[batchSize, inputFeatures]
+  -> Tensor dtype '[batchSize, outputFeatures]
+mlp MLP {..} = linear layer2 . relu . linear layer1 . relu . linear layer0
+
+model
+  :: MLP dtype inputFeatures outputFeatures hiddenFeatures
+  -> Tensor dtype '[batchSize, inputFeatures]
+  -> Tensor dtype '[batchSize, outputFeatures]
+model = (sigmoid .) . mlp
+
+foldLoop
+  :: forall a b m . (Num a, Enum a, Monad m) => b -> a -> (b -> a -> m b) -> m b
+foldLoop x count block = foldM block x ([1 .. count] :: [a])
+
+type BatchSize = 64
+type TestSize = 10000
+
+
+randomIndexes :: [Int]
+randomIndexes = map (`mod` natValI @I.DataDim) $ randoms seed
+  where
+    seed = mkStdGen 123
+
+
+toBackend :: String -> Tensor dtype shape -> Tensor dtype shape
+toBackend backend t =
+  case backend of
+    "CUDA" -> toCUDA t
+    _ -> toCPU t
+
+
+
+main = do
+  backend' <- try (getEnv "BACKEND") :: IO (Either SomeException String)
+  let backend = 
+        case backend' of
+          Right "CUDA" -> "CUDA"
+          _ -> "CPU"
+  let numIters = 10000
+  (trainingData, testData) <- I.initMnist
+  init    <- A.sample (MLPSpec :: MLPSpec 'D.Float I.DataDim I.ClassDim 128)
+  
+  
+  (trained,_)  <- foldLoop (init,randomIndexes) numIters $ \(state, idxs) i -> do
+    let (indexes,nextIndexes) = (take (natValI @I.DataDim) idxs, drop (natValI @I.DataDim) idxs)
+    let input           = toBackend backend $ I.getImages @BatchSize trainingData indexes 
+    let expected_output = toBackend backend $ I.getLabels @BatchSize trainingData indexes
+    let actual_output   = model state $ input
+    let loss            = mse_loss actual_output expected_output
+
+    let flat_parameters = A.flattenParameters state
+    let gradients       = A.grad (toDynamic loss) flat_parameters
+
+    when (i `mod` 250 == 0) (print loss)
+
+    new_flat_parameters <- mapM A.makeIndependent
+      $ A.sgd 1e-1 flat_parameters gradients
+    return $ (A.replaceParameters state new_flat_parameters, nextIndexes)
+
+  putStrLn $ "Learning loss:"
+  let test_data = toBackend backend $ I.getImages @TestSize testData [0..]
+      test_label = toBackend backend $ I.getLabels @TestSize testData [0..]
+  print $ mse_loss (model trained test_data) test_label
```

Comment:
while it'll probably still work with mse, nll_loss is probably more appropriate for the classification task https://github.com/pytorch/examples/blob/ee964a2eeb41e1712fe719b83645c79bcbd0ba1a/mnist/main.py#L34

Reply:
Likewise, mse is not a good choice, either. Luckily, typed negative log likelihood is already implemented 😅